go/tools/releaser/upgradedep_test.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"testing"
+
+	bzl "github.com/bazelbuild/buildtools/build"
+)
+
+func TestPatchItemParser(t *testing.T) {
+	tests := []struct {
+		expression []byte
+		result     string
+		error      string
+	}{
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			Label("//third_party:org_golang_x_tools-gazelle.patch")`),
+			result: "//third_party:org_golang_x_tools-gazelle.patch",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			"@io_bazel_rules_go//third_party:org_golang_x_tools-gazelle.patch"`),
+			result: "//third_party:org_golang_x_tools-gazelle.patch",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			"//third_party:org_golang_x_tools-gazelle.patch"`),
+			result: "//third_party:org_golang_x_tools-gazelle.patch",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			Label("@io_bazel_rules_go//third_party:org_golang_x_tools-gazelle.patch")`),
+			result: "@io_bazel_rules_go//third_party:org_golang_x_tools-gazelle.patch",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			NotLabel("//third_party:org_golang_x_tools-gazelle.patch")`),
+			result: "",
+			error:  `invalid patch function: "NotLabel"`,
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			NotLabel(True)`),
+			error: `invalid patch function: "NotLabel"`,
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			True`),
+			error: "not all patches are string literals or Label()",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			Label("//third_party:org_golang_x_tools-gazelle.patch", True)`),
+			error: "Label expr should have 1 argument, found 2",
+		},
+		{
+			expression: []byte(`# releaser:patch-cmd gazelle -repo_root . -go_prefix golang.org/x/tools -go_naming_convention import_alias
+			Label(True)`),
+			error: "Label expr does not contain a string literal",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%v", tt.expression), func(t *testing.T) {
+			patchExpr, err := bzl.Parse("repos.bzl", tt.expression)
+			if err != nil {
+				t.Fatalf(err.Error())
+			}
+
+			patchLabelStr, _, err := parsePatchesItem(patchExpr.Stmt[0])
+			if err != nil && err.Error() != tt.error {
```

Comment:
I prefer to break this into two test tables: `TestPatchItemParser_Success` and `TestPatchItemParser_Error`. That way we can keep the assertions exactly the same for each test table and avoid some branching logic in the test body. In general, we should avoid putting logic in the tests [1][2], because we have tests to cover production logic, but we don't have tests to cover test logic.

[1] https://learning.oreilly.com/library/view/software-engineering-at/9781492082781/ch12.html#donapostrophet_put_logic_in_tests
[2] https://learning.oreilly.com/library/view/software-engineering-at/9781492082781/ch12.html#tests_and_code_sharing_dampcomma_not_dr

Reply:
Totally fair, was mapping my original case (repos.bzl test file) into a single test, but obviously the landscape changed a little.